### PR TITLE
Features/hdev 549 display base apr on each pool in farm page

### PR DESCRIPTION
**Pull request by geonellov**
## Description of Changes
The changes here display the base apr of the pool on the Farm page, the base apr will display once the pool is added in DEPOSIT_TXHASH.

[Link to Jira Ticket](https://halodao.atlassian.net/browse/HDEV-549)

### Developer Checklist:

* [x] I have followed the guidelines in our Contributing document
* [x] This PR has a corresponding JIRA ticket
* [x] My branch conforms with our naming convention i.e. `feature/HDEV-XXX-description`
* [ ] All files have appropriate file headers and documentation
* [ ] I have written new tests for your core changes, as applicable
* [x] I have successfully ran tests locally

### Reviewers Checklist:
* [ ] Code is readable and understandable; any unclear parts have explanations 
* [ ] UI/UX changes match the corresponding figma/other design resources, if applicable
* [ ] I have successfully ran tests locally

### CODE REVIEW

File: src/halo-hooks/useRewards.ts [hunk 1]
```diff
@@ -307,3 +312,58 @@ export const useRewarderUSDPrice = (rewarderAddress: string | undefined) => {
...
+    // Calculate initial USD:HLP price during the first deposit tx of the pool
+    baseTokenInitialUsdPrice =
+      parseFloat(ethers.utils.formatUnits(txReceipt.logs[0].data, baseTokenInfo.decimals)) *
+      (await getInitialTokenUSDPrice(TOKEN_COINGECKO_NAME[baseToken.toLowerCase()], date))
+    if (baseTokenInfo.symbol === 'XSGD') {
+      quoteTokenInitialUsdPrice =
+        parseFloat(ethers.utils.formatUnits(txReceipt.logs[1].data, 6)) *
+        (await getInitialTokenUSDPrice(TOKEN_COINGECKO_NAME['usdc'], date))
+      hlpMintedValue = parseFloat(ethers.utils.formatUnits(txReceipt.logs[2].data, 18))
+    } else {
```
Comment (schystz) [inline src/halo-hooks/useRewards.ts]: @geonellov Q: so this else statement means other stables than XSGD (e.g. fxPHP, TGBP, TCAD), we use a different index for parsing the token amount and HLP minted from the tx log? Do we know the reason why?

File: src/utils/coingecko.ts [hunk 2]
```diff
@@ -29,3 +29,22 @@ export async function getTokensUSDPrice(by: GetPriceBy, addressesOrIds: string[]
     return prices
   }
 }
+
+export async function getInitialTokenUSDPrice(symbol: string, date: string) {
```
Comment (schystz) [inline src/utils/coingecko.ts]: Suggestion for a better name

```suggestion
export async function getTokenUSDPriceAtDate(symbol: string, date: string) {
```

File: src/constants/pools.ts [hunk 3]
```diff
@@ -128,3 +128,11 @@ const doubleEstimatePoolsKovan =
 export const DOUBLE_ESTIMATE_POOLS = doubleEstimatePools
 export const DOUBLE_ESTIMATE_POOLS_MATIC = doubleEstimatePoolsMatic
 export const DOUBLE_ESTIMATE_POOLS_KOVAN = doubleEstimatePoolsKovan
+
+// NOTE: add the pool address and its first deposit tx here for other chain
+export const DEPOSIT_TXHASH: Record<string, string> = {
```
Comment (schystz) [inline src/constants/pools.ts]: @geonellov so this means base apr on the UI will only show up on mainnet right?

I guess it's fine for now @ctverceles ?

**Review by schystz — COMMENTED** [review]
Good job overall @geonellov! The calculation looks accurate to me basing from https://halodao.atlassian.net/browse/HDEV-549?focusedCommentId=12778

UI looks good too!

I just have a couple of question + suggestion

File: src/constants/pools.ts (hunk 3 above)
Comment (chris247474) [reply src/constants/pools.ts]: @geonellov @schystz we want it to show up on all networks

File: src/components/Farm/FarmSummary.tsx [hunk 4]
```diff
@@ -61,86 +15,12 @@ const FarmSummary = ({ poolsInfo, tokenPrice }: FarmSummaryProps) => {
...
-              id="text-halo-earned"
-            >
-              {summary.haloEarned}
-            </TYPE.body>
-          </Row>
-        </EarnCard>
-      </Row>
-    </DashboardRow>
+    <div className="w-full flex flex-col md:flex-row space-y-2 md:space-y-0 md:space-x-2 pb-2">
+      <StakeCard title={t('totalPoolValue')} value="5000.03" />
```
Comment (chris247474) [inline src/components/Farm/FarmSummary.tsx:19]: @geonellov i see this is hardcoded. Looking at figma https://www.figma.com/file/lqpLD77mM2olJNrFzuD4Ul?node-id=7268:3943#148922421 it should be "Total Value Locked" which is the sum of all the USD values of `Total Pool Value` column

Or have we done this in another PR?

File: src/halo-hooks/useRewards.ts [hunk 5]
```diff
@@ -307,3 +312,58 @@ export const useRewarderUSDPrice = (rewarderAddress: string | undefined) => {
...
+
+    // Calculate current USD:HLP price
+    const [curveLiquidity, curveTotalSupply] = await Promise.all([
+      curveContract.liquidity(),
+      curveContract.totalSupply()
+    ])
+    currentUsdHlpPrice = parseFloat(formatEther(curveLiquidity.total_)) / parseFloat(formatEther(curveTotalSupply))
+
+    // Calculate the Base APR
+    setBaseApr((initialUsdHlpPrice / currentUsdHlpPrice - 1) * (365 / 190) * 100)
```
Comment (chris247474) [inline src/halo-hooks/useRewards.ts]: @geonellov we shouldn't hardcode 190 here. 

this should be `Base APR = ( ( initial_USD_HLP_price / current_USD_HLP_price ) - 1 ) * (365 days / # of days since first deposit tx) * 100 = % value`

so instead of `190` it should be `<# of days since first deposit tx>`

**Review by chris247474 — CHANGES_REQUESTED** [review]
@geonellov good start! left some comments, just need to adjust the hardcoded `190` in the calculation

Also, how long does it take for the calculation to show up? I tried to do a swap on UST:USDC to trigger a fee, but i dont see a number in the UI yet on UST:USDC farm page. maybe the price change between usd:hlp before and usd:hlp after is too small...?

In addition to my comments, @schystz how do you think we can unit test this `useGetBaseApr` function? i think i'd like to move it to a calculator/helper file then call it in `useRewards.ts` and the unit test. Does that make sense?

File: src/components/Farm/FarmSummary.tsx (hunk 4 above)
Comment (geonellov) [reply src/components/Farm/FarmSummary.tsx:19]: yep, it's done in another PR @ctverceles

Comment (schystz) [reply src/components/Farm/FarmSummary.tsx:19]: Oh yeah I was trying to find this too, because we are missing the $ on the UI

**Comment by schystz** [comment]
> In addition to my comments, @schystz how do you think we can unit test this `useGetBaseApr` function? i think i'd like to move it to a calculator/helper file then call it in `useRewards.ts` and the unit test. Does that make sense?

Yeah I think that will work @ctverceles.

There seem to be best practices for testing react components though like what is describe here: https://www.toptal.com/react/testing-react-hooks-tutorial but this is not something we tried or even explored yet.

File: src/halo-hooks/useRewards.ts (hunk 1 above)
Comment (geonellov) [reply src/halo-hooks/useRewards.ts]: @schystz resolved by just getting all the transfer logs

File: src/constants/pools.ts (hunk 3 above)
Comment (geonellov) [reply src/constants/pools.ts]: @schystz @ctverceles the pool base apr will show up on any chain only if its first deposit transaction is added in DEPOSIT_TXHASH constant.

**Comment by geonellov** [comment]
> Also, how long does it take for the calculation to show up? I tried to do a swap on UST:USDC to trigger a fee, but i dont see a number in the UI yet on UST:USDC farm page. maybe the price change between usd:hlp before and usd:hlp after is too small...?
@ctverceles yes, the price is too small, I computed it mannually and it is negative. also, reason why the UST:USDC base APR doesn't appear there is because it is not yet added to the DEPOSIT_TXHASH constant. I added it in my new commit.

File: src/constants/pools.ts (hunk 3 above)
Comment (chris247474) [reply src/constants/pools.ts]: ok @geonellov can we pls add this in the README? @tracyarciaga @schystz we'll need to take note of this so we add it for every network when we merge this

**Comment by chris247474** [comment]
> > In addition to my comments, @schystz how do you think we can unit test this `useGetBaseApr` function? i think i'd like to move it to a calculator/helper file then call it in `useRewards.ts` and the unit test. Does that make sense?
> 
> Yeah I think that will work @ctverceles.
> 
> There seem to be best practices for testing react components though like what is describe here: https://www.toptal.com/react/testing-react-hooks-tutorial but this is not something we tried or even explored yet.

@schystz @geonellov i was thinking more of a unit test just to check input vs output of the calculation. and we can move the calculation into its own `calculator.ts` function that we just import. this will make it easy for us to test

File: src/halo-hooks/useRewards.ts [hunk 6]
```diff
@@ -307,3 +312,62 @@ export const useRewarderUSDPrice = (rewarderAddress: string | undefined) => {
 
   return rewarderTokenUsdPrice[rewarderTokenAddress[0]]
 }
+
+export const useGetBaseApr = (poolAddress: string, tokenPair: string): number => {
+  const { account, library, chainId } = useActiveWeb3React()
+  const [baseAPR, setBaseApr] = useState(0)
+  const poolsWithBaseAPR = Object.keys(DEPOSIT_TXHASH)
+
+  const fetchBaseUsdHlpPrice = useCallback(async () => {
```
Comment (chris247474) [inline src/halo-hooks/useRewards.ts:331]: @geonellov should we move this calculation into its own calculator.ts function that we just import. this will make it easy for us to unit test different input/output

Comment (chris247474) [reply src/halo-hooks/useRewards.ts:331]: @geonellov i tried calculating TCAD:USDC base apr manually and this is what i got

[First deposit tx](https://etherscan.io/tx/0x6038346288968a18c766987b8a7811a2df17a25316edeb23800bb2125ba5902f) -> got the given for `initial_USD_HLP_price` from here

Curve.totalSupply = 108713.205624799189781903
Curve.liquidity = 109066.176879507049362387

```
initial_USD_HLP_price = (95.25 + 94.93)/189.503999999999999999 = 1.003567207

current_USD_HLP_price = 109066.176879507049362387 / 108,713.205624799189781903 = 1.0032468112

Base APR = ((1.003567207 / 1.0032468112) - 1) * (365/208) * 100 ) = 0.05604134577 %
```

**Review by chris247474 — CHANGES_REQUESTED** [review]
@geonellov @schystz 

we need to double check the `useBaseApr` function by:

- moving the calculation inside useBaseApr into a "helper" file like calculator.ts with function name calculateBaseApr(...)that you import into useReward.ts
- writing a test for calculateBaseApr(...) using the test input for xSGD:USDC (from [Jira](https://halodao.atlassian.net/browse/HDEV-549?focusedCommentId=12778))  and TCAD:USDC (from [my PR comment](https://github.com/HaloDAO/halodao-interface/pull/184/files#r846948894)
- assert that the 2 unit tests have roughly the same Base APR result as our manual calculations (within a certain threshold since the decimal places will a bit off sometimes)

File: src/tests/calculator.test.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
+import { calculateBaseApr } from '../utils/calculator'
+
+describe('Base APR Test', () => {
+  it('xSGD:USDC Mainnet', () => {
+    const baseTokenValue = 183024
```
Comment (chris247474) [inline src/tests/calculator.test.ts:7]: @geonellov can we put some comments explaining:

1. how we calculated this value
2. link to first deposit tx

File: src/tests/calculator.test.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+      hlpMintedValue,
+      totalCurveLiquidity,
+      totalCurveSupply,
+      noOfDaysSinceFirstDeposit
+    )
+    expect(actualBaseApr).toEqual(expectedBaseApr)
+  })
+
+  it('TCAD:USDC Mainnet', () => {
+    const baseTokenValue = 120
```
Comment (chris247474) [inline src/tests/calculator.test.ts:39]: @geonellov if we can put same comments as xSGD:USDC here

File: src/tests/calculator.test.ts [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+      hlpMintedValue,
+      totalCurveLiquidity,
+      totalCurveSupply,
+      noOfDaysSinceFirstDeposit
+    )
+    expect(expectedBaseApr).toEqual(actualBaseApr)
+  })
+
+  it('TGBP:USDC Mainnet', () => {
+    const baseTokenValue = 634.4908624191117
```
Comment (chris247474) [inline src/tests/calculator.test.ts:71]: @geonellov same comments here

File: src/tests/calculator.test.ts [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+      hlpMintedValue,
+      totalCurveLiquidity,
+      totalCurveSupply,
+      noOfDaysSinceFirstDeposit
+    )
+    expect(expectedBaseApr).toEqual(actualBaseApr)
+  })
+
+  it('fxPHP:USDC Mainnet', () => {
+    const baseTokenValue = 1003343.8102085856
```
Comment (chris247474) [inline src/tests/calculator.test.ts:103]: @geonellov same

File: src/tests/calculator.test.ts [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+      hlpMintedValue,
+      totalCurveLiquidity,
+      totalCurveSupply,
+      noOfDaysSinceFirstDeposit
+    )
+    expect(expectedBaseApr).toEqual(actualBaseApr)
+  })
+
+  it('UST:USDC Mainnet', () => {
+    const baseTokenValue = 9481.037924
```
Comment (chris247474) [inline src/tests/calculator.test.ts:135]: @geonellov same

**Review by chris247474 — CHANGES_REQUESTED** [review]
@geonellov thanks, looks better! but i can't figure out why TGBP:USDC is so high, it doesn't even have any swaps https://etherscan.io/txs?a=0x2ED09E2961D72659E4002ba8C2BaDfedC7db19B7

```
initial_USD_HLP_price = (974+881.92)/1,760.597934857799599999 = 1.0541418704
current_USD_HLP_price = 87141.993753048033038759 / 89487.665433794283801709 = 0.9737877654
Base APR = 14.4108743902
```

This doesn't seem believable so maybe for now lets not set the `DEPOSIT_TXHASH` var for TGBP:USDC until khidir and i figure this out. the rest of the pools seem good for now though! I think this is ok to merge to `develop` branch  after

- adding the comments in the `calculator.test.ts`
- removing `DEPOSIT_TXHASH` for TGBP:USDC
- adding `DEPOSIT_TXHASH` for the different pools of polygon and arbitrum

**Review by tracyarciaga — APPROVED** [review]
Test are passing, although is it possible to use the checksum format so that we can get rid of the warnings when running the test. 
@geonellov

File: src/halo-hooks/useRewards.ts [hunk 12]
```diff
@@ -307,3 +314,73 @@ export const useRewarderUSDPrice = (rewarderAddress: string | undefined) => {
...
+
+    // get the base token symbol and decimals
+    const baseTokenSymbol = tokenPair.split('/')[0]
+    const { decimals } = (haloTokenList[chainId as ChainId] as Token[]).filter(
+      token => token.symbol === baseTokenSymbol
+    )[0]
+
+    // get the transfer logs
+    const logs = txReceipt.logs.filter(
+      (value: any) => value.topics[0] === '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef'
```
Comment (schystz) [inline src/halo-hooks/useRewards.ts]: @geonellov I assume `0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef` is the "transfer" event hash? If so, can we put this in a constant file and reference it here as a const instead?

```suggestion
      (value: any) => value.topics[0] === TRANSFER_EVENT_HASH
```

**Review by schystz — APPROVED** [review]

File: src/constants/pools.ts [hunk 13]
```diff
@@ -133,9 +133,11 @@ export const DOUBLE_ESTIMATE_POOLS_KOVAN = doubleEstimatePoolsKovan
...
   '0xe15e50ff9d52bec41d53d3173f2ed40834d455f4': '0x6038346288968a18c766987b8a7811a2df17a25316edeb23800bb2125ba5902f', // TCAD:USDC mainnet
-  '0x2ed09e2961d72659e4002ba8c2badfedc7db19b7': '0xe1ba04f32987d607980ad89260942e16642ea74b23f1175a06dbc90ad0a30f28', // TGBP:USDC mainnet
+  // '0x2ed09e2961d72659e4002ba8c2badfedc7db19b7': '0xe1ba04f32987d607980ad89260942e16642ea74b23f1175a06dbc90ad0a30f28', // TGBP:USDC mainnet
   '0x20e1d8daf58358cf11be5616946e1df55f1ef8b0': '0x6dc4b1c61c00004c85ec953a0d00bd680bb42109977fb19bf4f4cfa7382910dc', // fxPHP:USDC mainnet
   '0x868084406449bda10a5bd556fb33cef5086b0797': '0xa605e6a3bf8c64b1a9f3e835dda1e61a77027789a6bde64dab43f91657d29081', // UST:USDC mainnet
-  '0x8123c64d6607412c7ac9e880f12245ef22558b14': '0x1baa65bca924e1a189ae68c0a57768d49f17e661de231566a2236383663f3fc7', // xSGD:USDC matic
-  '0x6156f030b877344470bac075a708d7e27602cc17': '0xafb8144ceea36323878efa628cbfcd41a01e91d931a61edb771d912fe2173920' // tagPHP:USDC matic
+  '0x8123c64d6607412c7ac9e880f12245ef22558b14': '0x1baa65bca924e1a189ae68c0a57768d49f17e661de231566a2236383663f3fc7' // xSGD:USDC matic
+  // '0x6156f030b877344470bac075a708d7e27602cc17': '', // tagPHP:USDC matic
```
Comment (chris247474) [inline src/constants/pools.ts]: @geonellov can we enable

- tagPHP:USDC matic
- fxPHP:USDC arbitrum?

**Review by chris247474 — CHANGES_REQUESTED** [review]
@geonellov can we enable the below pools as well?

File: src/constants/tokenLists/halo-tokenlist.ts [hunk 14]
```diff
@@ -203,3 +203,71 @@ export const haloAssimilators: { [chainId in ChainId]?: AssimilatorAddressMap }
...
+export const TOKEN_COINGECKO_NAME: Record<string, string> = {
+  USDC: 'usd-coin',
+  XSGD: 'xsgd',
+  TGBP: 'truegbp'
+}
+
+export type ChainLinkAddressMap = {
+  [token in TokenSymbol]: string
+}
+const mainNetToken: ChainLinkAddressMap = {
```
Comment (tracyarciaga) [inline src/constants/tokenLists/halo-tokenlist.ts]:
```suggestion
const mainnetToken: ChainLinkAddressMap = {
```

File: src/constants/tokenLists/halo-tokenlist.ts [hunk 15]
```diff
@@ -203,3 +203,71 @@ export const haloAssimilators: { [chainId in ChainId]?: AssimilatorAddressMap }
...
+  [TokenSymbol.XSGD]: ethers.constants.AddressZero,
+  [TokenSymbol.fxAUD]: ethers.constants.AddressZero,
+  [TokenSymbol.fxPHP]: '0xfF82AAF635645fD0bcc7b619C3F28004cDb58574',
+  [TokenSymbol.tagPHP]: ethers.constants.AddressZero,
+  [TokenSymbol.XIDR]: ethers.constants.AddressZero,
+  [TokenSymbol.fxAUD]: ethers.constants.AddressZero,
+  [TokenSymbol.UST]: ethers.constants.AddressZero
+}
+export const CHAINLINK_ORACLE: { [chainId in ChainId]?: ChainLinkAddressMap } = {
+  [ChainId.MAINNET]: mainNetToken,
```
Comment (tracyarciaga) [inline src/constants/tokenLists/halo-tokenlist.ts]:
```suggestion
  [ChainId.MAINNET]: mainnetToken,
```

File: src/constants/pools.ts [hunk 16]
```diff
@@ -128,3 +128,18 @@ const doubleEstimatePoolsKovan =
 export const DOUBLE_ESTIMATE_POOLS = doubleEstimatePools
 export const DOUBLE_ESTIMATE_POOLS_MATIC = doubleEstimatePoolsMatic
 export const DOUBLE_ESTIMATE_POOLS_KOVAN = doubleEstimatePoolsKovan
+
+// NOTE: add the pool address and its first deposit tx here for other chain
+export const DEPOSIT_TXHASH: Record<string, string> = {
```
Comment (schystz) [inline src/constants/pools.ts]: @geonellov can we group these by network just to be explicit? ...so something like:

```
export const DEPOSIT_TXHASH: { [chainId in ChainId]?: Record<string, string> } = {
  [ChainId.MAINNET]: {
    '0x64dcbdeb83e39f152b7faf83e5e5673faca0d42a': '0xead0a9717cb13feff2ee58d78786b7b2eac68eafe61809d7bfecf67986222ba8', // xSGD:USDC
    ...
  },
  [ChainId.MATIC]: {
    '0x8123c64d6607412c7ac9e880f12245ef22558b14': '0x1baa65bca924e1a189ae68c0a57768d49f17e661de231566a2236383663f3fc7', // xSGD:USDC
    ...
  },
  ...
}
```

File: src/constants/tokenLists/halo-tokenlist.ts [hunk 17]
```diff
@@ -203,3 +203,71 @@ export const haloAssimilators: { [chainId in ChainId]?: AssimilatorAddressMap }
...
+export const TOKEN_COINGECKO_NAME: Record<string, string> = {
+  USDC: 'usd-coin',
+  XSGD: 'xsgd',
+  TGBP: 'truegbp'
+}
+
+export type ChainLinkAddressMap = {
+  [token in TokenSymbol]: string
+}
+const mainnetToken: ChainLinkAddressMap = {
```
Comment (schystz) [inline src/constants/tokenLists/halo-tokenlist.ts]: How about renaming to `mainnetOracles`?

File: src/constants/tokenLists/halo-tokenlist.ts [hunk 18]
```diff
@@ -203,3 +203,71 @@ export const haloAssimilators: { [chainId in ChainId]?: AssimilatorAddressMap }
...
+  [TokenSymbol.TGBP]: ethers.constants.AddressZero,
+  [TokenSymbol.XSGD]: ethers.constants.AddressZero,
+  [TokenSymbol.fxAUD]: ethers.constants.AddressZero,
+  [TokenSymbol.fxPHP]: '0x9481e7ad8BE6BbB22A8B9F7B9fB7588d1df65DF6',
+  [TokenSymbol.tagPHP]: ethers.constants.AddressZero,
+  [TokenSymbol.XIDR]: ethers.constants.AddressZero,
+  [TokenSymbol.fxAUD]: ethers.constants.AddressZero,
+  [TokenSymbol.UST]: '0x8b6d9085f310396C6E4f0012783E9f850eaa8a82'
+}
+const polygonToken: ChainLinkAddressMap = {
```
Comment (schystz) [inline src/constants/tokenLists/halo-tokenlist.ts]: How about renaming to `polygonOracles`?

...line 252 to `arbOracles` as well

File: src/constants/pools.ts (hunk 3 above)
Comment (schystz) [reply src/constants/pools.ts]: > ok @geonellov can we pls add this in the README? @tracyarciaga @schystz we'll need to take note of this so we add it for every network when we merge this

Updated README

**Review by schystz — COMMENTED** [review]
@geonellov few minor comments

File: src/constants/pools.ts (hunk 13 above)
Comment (schystz) [reply src/constants/pools.ts]: @geonellov this is marked as resolved but I don't see the base APR for the pools above
